Fixes #10404 - Exporter: Allow user can export a custom value type other than string

jasonex7 opened the pull request:
Fixes #10404

Fix #10464

jasonex7 committed:
Fixes #10404 - Exporter: Allow user can export a custom value type other than string

jasonex7 committed:
Fixes #10404 - Fix TextExporter implementation

melloware commented on primefaces/src/main/java/org/primefaces/component/export/ColumnValue.java:
```diff
@@ -0,0 +1,64 @@
+    public boolean isCustomValue() {
+        return customValue != null;
+    }
+
```
Should probably implement equals() and hashcode() to be a good citizen

jasonex7 replied:
done.

melloware commented on primefaces/src/main/java/org/primefaces/component/export/ColumnValue.java:
```diff
@@ -0,0 +1,64 @@
+ */
+package org.primefaces.component.export;
+
+public class ColumnValue {
```
Should this be Serializable?

jasonex7 replied:
I don't think is necessary, because that object is not saved anywhere.

melloware commented on primefaces/src/main/java/org/primefaces/util/ExcelStylesManager.java:
```diff
@@ -251,6 +338,51 @@ private CellStyle createCurrencyStyle() {
+            dataFormat = wb.getCreationHelper().createDataFormat().getFormat(format);
+        }
+        else {
+            dataFormat = 14;
```
Where did this 14 come from?

jasonex7 replied:
It's a excel built-int format, documented in `org.apache.poi.ss.usermodel.BuiltinFormats`

melloware commented on primefaces/src/main/java/org/primefaces/util/ExcelStylesManager.java:
```diff
@@ -251,6 +338,51 @@ private CellStyle createCurrencyStyle() {
+            dataFormat = wb.getCreationHelper().createDataFormat().getFormat(format);
+        }
+        else {
+            dataFormat = 22;
```
Same here we should Javadoc where these constants came from?

jasonex7 replied:
I'm going to change the code to clarify where those values come from.

jasonex7 committed:
Fixes #10404 - Implements equals and hashCode

jasonex7 committed:
Fixes #10404 - Create and document constants for default date, time and datetime formats

jasonex7 committed:
Fixes #10404 - Fix typo

Rapster commented on primefaces/src/main/java/org/primefaces/component/export/ColumnValue.java:
```diff
@@ -0,0 +1,64 @@
+
+    public static ColumnValue fallbackValue(String value) {
+        return new ColumnValue(null, value);
+    }
```
Instead of having those 2 methods:

```java
public class ColumnValue {
        private final Object value;

        private ColumnValue(Object value) {
            this.value = value;
        }

        public static ColumnValue of(Object value) {
            return new ColumnValue(value);
        }

        public boolean isComplexValue() {
            return !isStringValue();
        }

        public boolean isStringValue() {
            return value instanceof String;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            ColumnValue that = (ColumnValue) o;
            return Objects.equals(value, that.value);
        }

        @Override
        public int hashCode() {
            return Objects.hash(value);
        }

        @Override
        public String toString() {
            return Objects.toString(value, Constants.EMPTY_STRING);
        }
    }
```

Makes you wonder how useful is this class... but it's better than working with Object type, and in the future maybe this class could be enriched? Also, toString() becomes quite handy here

jasonex7 replied:
I want to try the value as custom if the user set it in the attributes `exportValue` or `exportFunction` no matter if is String, LocalDate, BigDecimal or whatever type. So ExcelExporter could handle appropriately, please see my example in https://github.com/primefaces/primefaces/issues/10404#issue-1826639737

Rapster replied:
I understand that, but because you set exportValue or exportFunction doesn't necessarly mean it's a custom value, you might use that just because the column content is not supported by PF, example:

```xml
<p:column exportValue="#{car.color}>
    <div>
          <div>
              #{car.color}
          </div>
    </div>
</p:column>
```

If value is a String, you have to assume it's been properly formatted or formatted in a way that with stronglyTypedCells will try something smart. I understand you want to get rid of that tweak, but this is breaking change, for now stronglyTypedCells can simply be used as a fallback

melloware replied:
I agree with @Rapster. I actually do this in a few places where I may have custom format inside the column but the `exportValue` I use is a different field in the POJO that may be just a String that I want formatted by Strongly Typed Cells.

jasonex7 replied:
I've already change the name of the variable to "explicitlySetValue" instead of customValue to reflect what I want to mean. The problem with the actual version of PF is when I want to set some column as String and stronglyTypeCell tries to convert to number and strip characters and the excel column has some rows as text and some others as numbers, that is a bug for me. What do you suggest to solve that without stronglyTypeCell = false?

Rapster replied:
If I understand well, you have a String as input (which look like a number) and is exported as a number: that's what stronglyTypeCell does. Now if you don't want that, then simply disable stronglyTypeCell. There is no bug here so far

jasonex7 replied:
I'm sorry there is no bug, I mean it's a limitation of stronglyTypedCell. I don't want to disable for entire DataTable just for a single column that it is a String no matter how if looks like number, for example a national ID which starts with "0"

melloware replied:
So maybe on each column a new `exportStronglyTyped` boolean to allow it to be controlled per column?

Rapster replied:
@jasonex7 @melloware  you can create a map and make custom options per column

jasonex7 replied:
I think that new attribute in `<p:column>` could be called `exportOptions` and it will be a new interface similar to `org.primefaces.component.export.ExporterOptions` but for columns instead of whole DataTable. That will allow more config in the future. I don't like the map option because it would be too much generic.

Rapster replied:
Not sure what I had in mind at that time... Anyway, just to be clear, from what I understand you're trying to solve 2 issues no? First is, export raw values (as they are and no longer as String) and the other is to have, export options per column right?

jasonex7 replied:
I want primefaces receive the real value with exportValue or exportFunction properties in <p:colum> and handle that accordingly in exporter. But I understand that there will be a breaking change with String values set in that properties because now it's handle with stronglyTypeCell feature and with this PR as plain String. So in order to not introduce that breaking change, I suggest to add exportOptions property with stronglyTypedCell = true as default to preserve the current behavior.

Rapster replied:
> But I understand that there will be a breaking change with String values set in that properties

No, if a string is set, it's handled as it was before. Why this should be suddenly any different?

jasonex7 replied:
You're right, I see now what I have to do. I'm going to change this PR that in case stronglyTypedCell = true and the value set was a String then it behave like it was before, any other case is handle by this PR proposal. With that there will be no breaking change.

Rapster commented on primefaces/src/main/java/org/primefaces/util/ExcelStylesManager.java:
```diff
@@ -87,14 +95,44 @@ public static ExcelStylesManager createExcelStylesManager(Workbook wb, Locale lo
+        else {
+            updateCellFallbackValue(cell, value.getFallbackValue());
+        }
         applyColumnAlignments(column, cell);
```
If the exportValue is a String, it is considered as a custom value when it's not, this could cause regressions for some users. You should use value.isComplexType instead, and that's what updateCellCustomValue will do, only deal with non String value

jasonex7 replied:
That's not my intended use, I want to be capable of tell PrimeFaces I'm submitting a String value so ExcelExporter could treat like that and stop trying to guess what type is when stronglyTypedCells = true. I want to handle as customValue if the value was set with `exportValue` or `exportFunction` attribute in `<p:column>` despite of what type the value is. In all other cases, behave like before (ExcelExporter tries to figure what type is from the rendered value of the `<p:column>`)

Mabe I could rename the property as `explicitlySetValue` instead of `customValue`. I think that would show the real intention behind. What do you think?

Rapster replied:
First 2 lines will cause regression as explained before. Don't think custom/fallback have a lot of meaning here, it's more whether it's the raw value (which will require conversion) or string (where the conversion has already been made, but stronglyTypedCells  should still be taken account to remain backward compatible)

jasonex7 replied:
I've already change the name of the variable to "explicitlySetValue" instead of customValue to reflect what I want to mean. For now stronglyTypedCells is considered when the user don't use `exportValue` or `exportFunction` attributes of `<p:column>`, otherwise treats the value with type of the object. Maybe I could add another <p:column> attribute which permits disabling stronglyTypedCells per column basis.

jasonex7 replied:
I think that new attribute in `<p:column>` could be called `exportOptions` and it will be a new interface similar to `org.primefaces.component.export.ExporterOptions` but for columns instead of whole DataTable.

jasonex7 committed:
Fixes #10404 - Support custom values in header and footer also

jasonex7 committed:
Fixes #10404 - Fix support custom values in header and footer

jasonex7 committed:
Fixes #10404 - Use empty string instead of null in ColumnValue toString()

jasonex7 committed:
Fixes #10404 - ColumnValue: Rename customValue as explicitlySetValue to reveal real intent

The variable named "customValue" in the ColumnValue class was ambiguously named, making it harder to understand its purpose. This commit renames it to "explicitlySetValue", which more accurately describes its role: this value is set explicitly by the user, as opposed to being a default or fallback.

jasonex7 committed:
Fixes #10404 - Export Strings considering stronglyTypedCell option

Rapster commented on primefaces/src/main/java/org/primefaces/component/export/ExporterUtils.java:
```diff
@@ -185,28 +183,38 @@ public static String getComponentFacetValue(FacesContext context, UIComponent pa
+        ColumnValue columnValue = ColumnValue.EMPTY_VALUE;
         if (columnType == TableExporter.ColumnType.HEADER) {
-            textValue = Optional.ofNullable(column.getExportHeaderValue()).orElseGet(column::getHeaderText);
+            if (column.getExportHeaderValue() != null) {
```
Keep the original code, you can do the same in 2 lines

Rapster commented on primefaces/src/main/java/org/primefaces/component/export/ExporterUtils.java:
```diff
@@ -185,28 +183,38 @@ public static String getComponentFacetValue(FacesContext context, UIComponent pa
         }
         else if (columnType == TableExporter.ColumnType.FOOTER) {
-            textValue =  Optional.ofNullable(column.getExportFooterValue()).orElseGet(column::getFooterText);
+            if (column.getExportFooterValue() != null) {
```
Keep the original code, you can do the same in 2 lines

Rapster commented on primefaces/src/main/java/org/primefaces/component/export/ExporterUtils.java:
```diff
@@ -185,28 +183,38 @@ public static String getComponentFacetValue(FacesContext context, UIComponent pa
         UIComponent facet = column.getFacet(columnType.facet());
-        if (LangUtils.isBlank(textValue) && ComponentUtils.shouldRenderFacet(facet)) {
-            textValue = getComponentValue(context, facet);
+        if (LangUtils.isBlank(columnValue.toString()) && ComponentUtils.shouldRenderFacet(facet)) {
```
You could introduce `ColumnValue#isBlank` since ColumnValue should never be null

Rapster reviewed (changes requested):
Looks much better now, I left few notes

melloware committed:
Code review comments

melloware committed:
Update ExporterUtils.java
